### PR TITLE
include/denc: improvements

### PR DESCRIPTION
**Pull request by adamemerson**
Collapse duplicated code between all containers.

(I know about the test failures, I'll run through those.)

### CODE REVIEW

File: src/include/denc.h [hunk 1]
```diff
@@ -156,6 +158,22 @@ struct denc_traits {
...
+  ===================
+
+  A class with a constructor like
+
+    class Foo {
+      ...
+      Foo(buffer::ptr::iterator& p, uint64_t features = 0);
+    }
+
+  Will have be deconde constructble. Instances of this class can be
```
Comment (liewegas) [inline src/include/denc.h]: typo

File: src/include/denc.h [hunk 2]
```diff
@@ -156,6 +158,22 @@ struct denc_traits {
...
+  efficiently decoded into containers.
+
+  Since decode-constructors are new, I require all of them to accept a
+  features argument, to avoid the combinatorial explosion.
```
Comment (liewegas) [inline src/include/denc.h]: We need to be very careful with this.  We generally rely on being able to transition a class from unfeatured to featured and get compile errors on all encoding paths.  If callers for this pass in feature fields they aren't explicitly passed (e.g., from a features=0 arg) we may silently fill in bogus features.  (Hopefully we haven't already done this elsewhere!)

**Comment by liewegas** [comment]
Several build errors.. see https://jenkins.ceph.com/job/ceph-pull-requests/16112/

**Comment by liewegas** [comment]
The ctor changes are great!

**Comment by adamemerson** [comment]
@liewegas Take another look at this? I plan to redo the constructor stuff on top of it to see if I can get it to work under Trusty Tahr. I don't know if you want to pull this in as is since it's a logically separate change, or hold off until I have the whole thing done.

**Comment by liewegas** [comment]
Let's keep them separate

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
retest this please

File: src/test/test_denc.cc [hunk 3]
```diff
@@ -518,9 +518,24 @@ TEST(denc, array)
...
+}
+
+TEST(denc, tuple)
+{
   {
-    cout << "std::array<legacy_t, 3>" << std::endl;
-    std::array<legacy_t, 2> s = { legacy_t(1), legacy_t(2) };
-    test_encode_decode(s);
+    cout << "std::tuple<uint64_t, uint32_t>" << std::endl;
+    std::tuple<uint64_t, uint32_t> s = { 100, 97 };
```
Comment (tchaikov) [inline src/test/test_denc.cc]: the jenkins fails. so might want to put something like 
```c++
std::tuple<uint64_t, uint32_t> s = { 100U, 97U };
```
instead

File: src/test/test_denc.cc [hunk 4]
```diff
@@ -518,9 +518,24 @@ TEST(denc, array)
...
+    counts.reset();
+    test_denc(s);
+  }
+  {
+    cout << "std::tuple<std::string, uint3_t>" << std::endl;
+    std::tuple<std::string, uint32_t> s = { "foo", 97 };
```
Comment (tchaikov) [inline src/test/test_denc.cc]: ditto.

File: src/test/test_denc.cc [hunk 5]
```diff
@@ -518,9 +518,24 @@ TEST(denc, array)
...
+    test_denc(s);
+  }
+  {
+    cout << "std::tuple<std::string, std::set<uint32_t>>" << std::endl;
+    std::tuple<std::string, std::set<uint32_t>> s = { "bar", {1, 2, 3} };
```
Comment (tchaikov) [inline src/test/test_denc.cc]: ditto.

**Comment by adamemerson** [comment]
@liewegas All right! The tests are hammered into submission.

**Comment by liewegas** [comment]
passed rados suite test, http://pulpito.ceph.com/sage-2016-12-29_20:50:13-rados-wip-sage-testing---basic-smithi/

**Comment by adamemerson** [comment]
@liewegas Do you mean it didn't pass? I see a couple erros but they look like clock skews and problems with the testing infrastructure.

**Comment by liewegas** [comment]
I meant it passed; failures do not appear related.

This all looks fine to me.  I'm not sure that boost::optional and tuple are that useful, since we aren't explicitly encoding those types, but it's certainly harmless to support them.  With optional, for example, it is probably just as likely that we will use a shared flags field to indicate the presence of optional fields.  But it does happen...

**Comment by liewegas** [comment]
retest this please

**Pull request merged**